File: 2023/day04-1.ts
type Scratchcard = {
  id: string;
  numbers: number[];
  winning: number[];
};

function solution(content: string) {
  return content
    .split("\n")
    .map(parse)
    .map(winning)
    .filter((item) => item > 0)
    .map((item) => Math.pow(2, item - 1))
    .reduce((acc, cur) => acc + cur, 0);
}

function winning(card: Scratchcard) {
  return card.winning.reduce(
    (acc, cur) => acc + (card.numbers.includes(cur) ? 1 : 0),
    0
  );
}

function parse(line: string): Scratchcard {
  const match = line.match(/^Card\s*(\d+):\s*([\d\s]+)\s*\|\s*([\d\s]+)$/);
  if (!match) throw new Error(`Invalid line ${line}`);
  return {
    id: match[1],
    numbers: match[2]
      .replace(/\s+/, " ")
      .split(" ")
      .filter((item) => item)
      .map(Number),
    winning: match[3]
      .replace(/\s+/, " ")
      .split(" ")
      .filter((item) => item)
      .map(Number),
  };
}

Deno.readTextFile("day04.dat")
  .then((content) => content.trim())
  .then(solution)
  .then(console.log);

File: 2023/day04_1.py
# pylint: disable=missing-module-docstring, missing-function-docstring


import re
from typing import Any

type Scratchcard = tuple[list[int], list[int]]


def solution(content: list[str]) -> Any:
    return sum(
        2 ** (n - 1) for n in (winning(parse(line)) for line in content) if n > 0
    )


def winning(card: Scratchcard):
    nums, wnums = card
    return len([True for n in nums if n in wnums])


def parse(line: str) -> Scratchcard:
    match = re.match(r"Card\s*\d+:\s*([\d\s]+)\s*\|\s*([\d\s]+)", line)
    if match is None:
        raise ValueError(f"Invalid line: {line}")
    return (
        [int(x) for x in re.sub(r"\s+", " ", match.group(1)).strip().split(" ")],
        [int(x) for x in re.sub(r"\s+", " ", match.group(2)).strip().split(" ")],
    )


def main() -> None:
    with open("day04.dat", "r", encoding="utf8") as file:
        print(solution(file.readlines()))


if __name__ == "__main__":
    main()
